// @jorge/lib-di/index.ts
import {IContainer} from 'container-ioc';

import { Container, LifeTime, Injectable } from "container-ioc";
import { IModule } from './lib/interfaces';
import { IServiceProvider, ServiceProvider } from "./lib/serviceprovider";
import { ContainerConfiguration } from './lib/configuration';
import { TServiceProvider } from './lib/symbols';

export * from "./lib/configuration"
export * from "./lib/symbols"
export * from "./lib/interfaces"
export * from "./lib/serviceprovider"

let containerConfiguration: ContainerConfiguration = new ContainerConfiguration();

let globalContainer: Container = new Container();

export function getGlobalServiceProvider() : IServiceProvider
{
    return <IServiceProvider>globalContainer.resolve(TServiceProvider);
}

export function registerModule(m: IModule) {
  m.registerServices(containerConfiguration);
}

export function createChildContainer() : IContainer {
  return globalContainer.createScope();
}

export function createChildServiceProvider(parentContainer?: IContainer) : IServiceProvider
{
  if (parentContainer == null)
  {
    let childContainer: IContainer = globalContainer.createScope();
    return new ServiceProvider(childContainer);  
  }

  let childContainer: IContainer = parentContainer.createScope();
  return new ServiceProvider(childContainer);
}

export function initializeDependencyInjection()
{
  containerConfiguration.configurationContainer(globalContainer);
}

// register service provider as a singleton within the container
globalContainer.register([
  {
    token: TServiceProvider,
    useFactory: () => new ServiceProvider(globalContainer),
    lifeTime: LifeTime.Persistent
  }
]);
