// @jorge/serverapp/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var express = require("express");
var path = require("path");
require("reflect-metadata");
var lib_webapphost_1 = require("@jorge/lib-webapphost");
var di = require("@jorge/lib-di");
var Server = /** @class */ (function () {
    /**
     *
     */
    function Server() {
        this.serviceProvider = di.getGlobalServiceProvider();
        this.host = this.serviceProvider.getService(lib_webapphost_1.TApplication);
        this.host.baseDirectory = path.join(__dirname, "../");
        this.host.addRoute("/", new DefaultRoutes());
    }
    /**
     * start
     */
    Server.prototype.start = function (port) {
        this.host.initialize();
        this.host.start(port);
    };
    return Server;
}());
var DefaultRoutes = /** @class */ (function () {
    function DefaultRoutes() {
        this.router = express.Router();
        this.router.get("/", function (req, res) {
            var container = Reflect.getMetadata("container", req);
            var value = container.getService(Symbol.for("Name"));
            res.render("index", { title: "Express [" + value + "]" });
        });
    }
    DefaultRoutes.prototype.getExpressRouter = function () {
        return this.router;
    };
    return DefaultRoutes;
}());
var server = new Server();
server.start(3000);
